### PR TITLE
better error message for trailing doc comment in traits (issue #56766)

### LINKED ISSUE #56766: Blank document comment in trait raises error different from with that in function.
If I use document comment that doesn't document anything in function,

```Rust
fn test(){
    println!("Hello world");
    ///
}
```

it will raise error:
```
error[E0585]: found a documentation comment that doesn't document anything.
```

But if I use this in trait, the error confuses me.

```Rust
trait User{
    fn test();
    ///
}
fn main() {
    println!("Hello world");
}
```

```
error: expected one of `async`, `const`, `extern`, `fn`, `type`, or `unsafe`, found `}`
 --> src/main.rs:4:1
  |
3 |     ///
  |        - expected one of `async`, `const`, `extern`, `fn`, `type`, or `unsafe` here
4 | }
  | ^ unexpected token
```

### PR DESCRIPTION
**Pull request by lcnr**
closes #56766 
r? @estebank

### CODE REVIEW

**Review by estebank — APPROVED** [review]
Did you mean to modify `src/libbacktrace` in your PR? If you didn't, could you rewrite your history to not include it? Also, could you squash your commits?

Other than that it looks good to me.

**Comment by lcnr** [comment]
@estebank Done. It seems like I somehow automatically created `src/libbacktrace` as the folder is was completely new. :thinking:

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -5940,6 +5940,14 @@ impl<'a> Parser<'a> {
             self.expect(&token::OpenDelim(token::Brace))?;
             let mut trait_items = vec![];
             while !self.eat(&token::CloseDelim(token::Brace)) {
+                if let token::DocComment(_) = self.token {
+                    if self.look_ahead(1, |tok| tok == &token::Token::CloseDelim(token::Brace)) {
+                        self.span_fatal_err(self.span, Error::UselessDocComment).emit();
```
Comment (estebank) [inline src/libsyntax/parse/parser.rs]: It'd be nice if you used `struct_span_err(...).emit()` instead, and then consume the doc (as you're doing here with `self.bump()`. That way the parser will "succeed" and continue to type checking in the face of this.

File: src/test/ui/useless-doc-comment/trait-impl.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+trait Foo {
+    ///
+    fn foo(&self);
+    /// I am not documenting anything
+    //~^ ERROR: found a documentation comment that doesn't document anything [E0585]
+}
+
+fn main() {
+
```
Comment (estebank) [inline src/test/ui/useless-doc-comment/trait-impl.rs]: Add a `let _: usize ="";` line to make sure the compiler recovers from the incorrect doc comment.

**Review by estebank — COMMENTED** [review]
@lcnr sorry for the delay in reviewing this. You will need to rebase against master and fix the merge conflict. I have one nitpick, but otherwise, the change looks great! Thank you!

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (lcnr) [reply src/libsyntax/parse/parser.rs]: I tried doing that, but I was unable to find a way to emit an `Error` without using `span_fatal_err`. All other methods of `Parser`currently use a `&str` instead. I might look deeper into the diagnostics system once I have the time. Might take a while though :cry:

**Comment by TimNN** [comment]
Ping from triage @lcnr: What is the status of this PR?

**Comment by lcnr** [comment]
I currently don't have much time and will probably want to improve the error functions of `Parser` before finishing this pull request. I would probably keep this open for a few more week to not forget this.

**Comment by lcnr** [comment]
I don't think that I can finish this in the near future.

**Pull request closed without merging**